+neurostim/+plugins/mcc.m

```diff
@@ -126,25 +127,42 @@ function afterExperiment(o)
         function digitalOut(o,channel,value,varargin)
             % digitalOut(o,channel,value [,duration])
             % Output the value to the digital channel
-            % o.digitalOut(0,unit8(2)) will write '2' to port A
-            % o.digitalOut(3,false) will set bit #3 to false.
+            % There are two modes here:
+            % 1) if VALUE is uint8, a full byte is written to port CHANNEL
+            % 2) if VALUE is logical, the single bit is set on line CHANNEL
+            % Currently Port0/A is hard-coded as input and Port1/B as
+            % output, so CHANNEL should be 1 (port B) or 9-16 (lines 1-8 in port B)
+            %
+            % o.digitalOut(1,uint8(2)) will write '2' to port B
+            % o.digitalOut(11,false) will set bit #3 on port B to false.
             if isa(value,'uint8') && ismember(channel ,[0 1])
                 % Writing a full byte to port A (channel 0) or  B (1)
                 DaqDOut(o.daq,channel,value);
+                o.dOutOld(channel+1) = value;
             elseif islogical(value)
                 % Set a single bit
                 % First get the current values of both Ports A & B;
-                current = DaqDIn(o.daq);
-                port = (channel>8)+1; %Determine which port the bit number belongs to.
-                current = current(port); %Retrieve current value of the port 
-                newValue = bitset(current,mod(channel,8),value); 
-                DaqDOut(o.daq,port-1,newValue);
+                % current = DaqDIn(o.daq);
+                % port = (channel>8)+1; %Determine which port the bit number belongs to.
+                % current = current(port); %Retrieve current value of the port 
+                % DaqDOut(o.daq,port-1,newValue);
+                
+                % using DaqDIn is sloow (i.e. framedrop slow). Simpler to just save the value
+                % note that this allows setting of multiple lines on a single port simultaneously
+                % channel(a) defines the physical line and value(a) the
+                % logical value it should be set to
+                newValue = o.dOutOld((channel(1)>8)+1); 
+                for a = 1:length(value)
+                    newValue = bitset(newValue,mod(channel(a),8),value(a)); 
+                end
+                DaqDOut(o.daq,channel(1)>8, newValue);
+                o.dOutOld((channel(1)>8)+1) = newValue;
                 
                 if size(varargin) == 1 
                     duration = varargin{1};
                     % timer function may override other functions when time is met
                     % and could cause problems for time-critical tasks
-                    o.timer = timer('StartDelay',duration/1000,'TimerFcn',@(~,~) outputToggle(o,channel,current)); %#ok<CPROPLC>
+                    o.timer = timer('StartDelay',duration/1000,'TimerFcn',@(~,~) outputToggle(o,channel(1),newValue)); %#ok<CPROPLC>

```

Comment:
why channel(1) - better to check that it is a scaler right at the start and then use that variable as is?  (same comment fo the code above where channel(1) is used.

Reply:
Thanks for the comment. One of motivations of the current fix is to allow multiple lines to be set at once. Thus the channels can be a vector to allow the following usage: digitalOut([line1 line2], [true false]). The lines must belong to the same port, though.

Apart from the channel dimensionality, I noticed that the 3rd input to digitalOut does not work as intended. I will fix this in another pull request.